File: projects/restdocs/webtestclient/src/main/java/com/gmail/unmacaque/spring/restdocs/webtestclient/config/RouteConfiguration.java
package com.gmail.unmacaque.spring.restdocs.webtestclient.config;

import com.gmail.unmacaque.spring.restdocs.webtestclient.repository.MessageRepository;
import com.gmail.unmacaque.spring.restdocs.webtestclient.web.MessageRouteHandler;
import org.jspecify.annotations.NonNull;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration(proxyBeanMethods = false)
public class RouteConfiguration {

	@Bean
	MessageRouteHandler routeHandler(MessageRepository messageRepository) {
		return new MessageRouteHandler(messageRepository);
	}

	@Bean
	RouterFunction<@NonNull ServerResponse> routerFunction(MessageRouteHandler handler) {
		return route()
				.GET("/messages", handler::getMessages)
				.POST("/messages", handler::postMessage)
				.build();
	}
}

File: spring-kotlin/src/main/kotlin/com/gmail/unmacaque/spring/config/RouteConfiguration.kt
package com.gmail.unmacaque.spring.config

import com.gmail.unmacaque.spring.domain.MessageRepository
import com.gmail.unmacaque.spring.web.MessageHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.servlet.function.router

@Configuration
class RouteConfiguration {

    @Bean
    fun routeHandler(repository: MessageRepository) = MessageHandler(repository)

    @Bean
    fun routerFunction(messageHandler: MessageHandler) =
        router {
            "/messages".nest {
                GET("/", messageHandler::getMessages)
                GET("/{id}", messageHandler::getMessage)
                POST("/", messageHandler::postMessage)
            }
        }

}
